### ISSUE TITLE
Use active function relu ,and the out is all nan.

### ISSUE THREAD

**Issue by Chenly1993** [tiny-dnn/tiny-dnn#696]
Now I want  to learn a vector from a picture. Because the output is a vector, I use fit() to train and the loss function is cross_entropy. And I need to use relu as the active function for all layers. But the output is all not an number. I don't know how to solve the problem. Who can help me？Thanks a lot！
![Uploading 图片1.png…]()

**Comment by Chenly1993** [comment]
![1](https://cloud.githubusercontent.com/assets/28524540/25793038/c655340a-33fc-11e7-97c8-8b6cb9f284bb.png)

**Comment by Chenly1993** [comment]
![2](https://cloud.githubusercontent.com/assets/28524540/25793063/e35a6b10-33fc-11e7-8a3d-831e2d0fcfa4.png)

**Comment by Randl** [comment]
Can you provide some code?

**Comment by Chenly1993** [comment]
`#include <iostream> 
#include <stdlib.h>
#include <stdio.h>
#include <string>
#include <algorithm>
#include <fstream>
#include "tiny_dnn/tiny_dnn.h"
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <boost/foreach.hpp>
#include <boost/filesystem.hpp>
#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>


using namespace boost::filesystem;

const int faceNum = 640;
tiny_dnn::dropout_layer dropout(2000, 0.5);
tiny_dnn::dropout_layer dropout1(256, 0.5);

static void construct_net(tiny_dnn::network<tiny_dnn::sequential>& nn)
{

	// by default will use backend_t::tiny_dnn unless you compiled
	// with -DUSE_AVX=ON and your device supports AVX intrinsics
	tiny_dnn::core::backend_t backend_type = tiny_dnn::core::default_engine();

	try {
		// construct nets: C: convolution; S: sub-sampling; F: fully connected
		nn << tiny_dnn::convolutional_layer<tiny_dnn::activation::tan_h>(64, 64, 7, 1, 32,  // C1, 1@64x64-in, 32@58x58-out
			tiny_dnn::padding::valid, true, 1, 1, backend_type)
			<< tiny_dnn::max_pooling_layer<tiny_dnn::activation::tan_h>(58, 58, 32, 2)   // S2, 32@58x58-in, 32@29x29-out
			<< tiny_dnn::convolutional_layer<tiny_dnn::activation::tan_h>(29, 29, 7, 32, 64, // C3, 32@29x29-in, 64@23x23-out
				tiny_dnn::padding::valid, true, 1, 1, backend_type)
			<< tiny_dnn::max_pooling_layer<tiny_dnn::activation::tan_h>(23, 23, 64, 2)  // S4, 64@23x23-in, 64@11x11-out
			<< tiny_dnn::fully_connected_layer<tiny_dnn::activation::tan_h>(7744, 2000,        // F5, 7744-in, 2000-out
				true, backend_type)
			<< dropout
			<< tiny_dnn::fully_connected_layer<tiny_dnn::activation::sigmoid>(2000, 256,        // F6, 2000-in, 256-out
				true, backend_type);
	}
	catch (const tiny_dnn::nn_error& e) {
		std::cout << e.what();
	}
	for (int i = 0; i < nn.depth(); i++) {
		std::cout << "#layer:" << i << "\n";
		std::cout << "layer type:" << nn[i]->layer_type() << "\n";
		std::cout << "input:" << nn[i]->in_data_size() << "(" << nn[i]->in_data_shape() << ")\n";
		std::cout << "output:" << nn[i]->out_data_size() << "(" << nn[i]->out_data_shape() << ")\n";
	}
}

static void convert_image(const std::string& imagefilename, double minv, double maxv, int w, int h, tiny_dnn::vec_t& data)
{
	tiny_dnn::image<> img(imagefilename, tiny_dnn::image_type::grayscale);
	tiny_dnn::image<> resized = resize_image(img, w, h);

	// mnist dataset is "white on black", so negate required
	std::transform(resized.begin(), resized.end(), std::back_inserter(data),
		[=](uint8_t c) { return (255 - c) * (maxv - minv) / 255.0 + minv; });
}


static void train_lenet(const std::string& data_dir_path)
{
	// specify loss-function and learning strategy
	tiny_dnn::network<tiny_dnn::sequential> nn;
	tiny_dnn::adagrad optimizer;

	construct_net(nn);

	std::cout << "load models..." << std::endl;

	//Load Yale Face
	std::vector<tiny_dnn::vec_t> train_images, test_images;
	std::vector<tiny_dnn::vec_t> train_codes, test_codes;
	std::vector<std::string> code;

	ifstream fin("E:\\MEB_code.txt");
	std::string  s;
	while (getline(fin, s))
	{
		code.push_back(s);
	}
	fin.clear();
	fin.close();


	for (int i = 1; i < 39; i++) {
		std::string str = std::to_string(i);
		std::string filename;
		str = data_dir_path + "/yaleB" + str + "/";

		tiny_dnn::vec_t vec;
		tiny_dnn::vec_t out;

		std::string c = code[i - 1];
		std::vector<char> Mcode;
		for (int n = 0; n < 256; n++) {
			Mcode.push_back(c.at(n));
		}
		
		for (int n1 = 0; n1 < 256; n1++) {
			float m = Mcode[n1]-'0';
			out.emplace_back(m);
		}
		
		for (int j = 0; j < faceNum; j++) {
			// convert imagefile to vec_t
			filename=str+ std::to_string(j+1)+".jpg";
			tiny_dnn::image<> gray_img(filename, tiny_dnn::image_type::grayscale);
			vec = gray_img.to_vec();
			//convert_image(filename, -1.0, 1.0, 64, 64, vec);
			int n2 = j + 1;
			if (n2 % 64 == 0)
			{
				test_images.push_back(vec);
				test_codes.emplace_back(out);
				vec.clear();
				continue;
			}
			train_images.push_back(vec);
			train_codes.emplace_back(out);
			vec.clear();
		}
		Mcode.clear();
		out.clear();
		std::cout << "读入第" << i << "个人的图片" << std::endl;
	}



	std::cout << "start training" << std::endl;
	tiny_dnn::progress_display disp(static_cast<unsigned long>(train_images.size()));
	tiny_dnn::timer t;
	int minibatch_size = 64;
	int num_epochs = 20;
	optimizer.alpha = 0.1;
	int count = 1;

	// create callback
	auto on_enumerate_epoch = [&]() {

		std::cout << t.elapsed() << "s elapsed." << std::endl;
		//tiny_dnn::result res = nn.test(test_images, test_codes);
		double loss = nn.get_loss<tiny_dnn::cross_entropy>(test_images, test_codes);
		std::cout <<"loss:" <<loss<< std::endl;
		dropout.set_context(tiny_dnn::net_phase::train);
		optimizer.alpha *= 0.90;  // decay learning rate
		optimizer.alpha = std::max((tiny_dnn::float_t)0.00001, optimizer.alpha);
		disp.restart(static_cast<unsigned long>(train_images.size()));
		t.restart();
	};

	auto on_enumerate_minibatch = [&]() {
		disp += minibatch_size;
	};

	// training
	nn.fit<tiny_dnn::cross_entropy>(optimizer, train_images, train_codes, minibatch_size,
		num_epochs, on_enumerate_minibatch, on_enumerate_epoch);

	std::cout << "end training." << std::endl;


	// save network model & trained weights
	nn.save("LeNet-model");

	code.clear();
}


// rescale output to 0-100
template <typename Activation>
static double rescale(double x)
{
	Activation a;
	return 100.0 * (x - a.scale().first) / (a.scale().second - a.scale().first);
}


int test_dnn_yale_train()
{
	std::string data_dir_path = "E:/database/Yale2_2_T";
	train_lenet(data_dir_path);
	return 0;
}


int main()
{
	int train = test_dnn_yale_train();
	if (train == 0) fprintf(stderr, "train success\n");
	else fprintf(stderr, "train fail\n");
	system("pause");
	return 0;
}
`

**Comment by Chenly1993** [comment]
This is all my code!

**Comment by Chenly1993** [comment]
I need to use relu. And if I use tanh, the loss is still **nan**.

**Comment by bhack** [comment]
@Randl If it is not a bug can we move on stackoverflow? Cause here support replies and knowledge base will be lost.

**Comment by bhack** [comment]
We have already a specific tag to feed http://stackoverflow.com/questions/tagged/tiny-dnn
